src/java.base/share/classes/java/nio/file/FileSystems.java

```diff
@@ -112,11 +113,9 @@ private static FileSystemProvider getDefaultProvider() {
             if (propValue != null) {
                 for (String cn: propValue.split(",")) {
                     try {
-                        Class<?> c = Class
-                            .forName(cn, true, ClassLoader.getSystemClassLoader());
-                        Constructor<?> ctor = c
-                            .getDeclaredConstructor(FileSystemProvider.class);
-                        provider = (FileSystemProvider)ctor.newInstance(provider);
+                        Class<?> c = Class.forName(cn, true, ClassLoaders.appClassLoader());
```

Comment:
This breaks the case where a custom system `ClassLoader` doesn’t depend on a custom default `FileSystemProvider` but the latter depends on the custom system `ClassLoader`, so this should call `ClassLoaders​::appClassLoader()` only when `VM.isBooted()` is `false`.

Reply:
A custom default file system provider can't depend on a custom system ClassLoader. Think of access to the default file system as something that works at a much lower level that a class loader.